### PR TITLE
Improve algo.SPpaths single shortest path performance

### LINKED ISSUE #2162: Improve algo.SPpaths single path performance

### PR DESCRIPTION
**Pull request by swilly22**
Resolve #2162

## Release Notes

* **Performance / Behavior**
  * Improved constrained weighted path search with reused adjacency/edge lookup and a structural reachability pre-check to tighten bounds.
  * Added an exact Dijkstra fast path for eligible single shortest-path queries to return correct minimum path weights.
* **Bug Fixes**
  * Prevented duplicate/degenerate edge candidates from corrupting k-minimal results.
  * Ensured unreachable source/target pairs return zero results and refined source=target handling to respect minimum length semantics.
* **Tests**
  * Added Dijkstra-dispatch correctness checks (including unreachable cases) and expanded coverage for multiple graph shapes and edge cases.

### CODE REVIEW

File: src/procedures/proc_sp_paths.c [hunk 1]
```diff
@@ -437,6 +440,441 @@ static inline SIValue _get_value_or_default
...
+// maxLen hops, in which case no path can exist regardless of maxCost
+// either, and the caller can skip the exhaustive search entirely.
+static void _find_bound_path
+(
+	SinglePairCtx *ctx,
+	bool *exists,
+	double *out_weight,
+	double *out_cost
+) {
+	*exists = false;
```
Comment (DvirDukhan) [inline src/procedures/proc_sp_paths.c:488]: should this be after the check of `src_id == dat_id`?

File: tests/flow/test_path_algorithms.py [hunk 2]
```diff
@@ -508,3 +510,263 @@ def test08_fractional_weights(self):
...
+        results = [
+            self.sp_query(self.n, self.m, ["E"], 3, None, 1, None),
+            self.sp_query(self.n, self.m, None, 3, None, 1, None)
+        ]
+
+        for result in results:
+            self.env.assertEquals(len(result.result_set), 1)
+            self.env.assertAlmostEqual(result.result_set[0][1], min_weight, delta=1e-9)
+
+    def test11_sp_unreachable(self):
```
Comment (DvirDukhan) [inline tests/flow/test_path_algorithms.py]: test 10

File: src/procedures/proc_sp_paths.c [hunk 3]
```diff
@@ -683,20 +1135,76 @@ static ProcedureResult Proc_SPpathsInvoke
...
+	_find_bound_path(single_pair_ctx, &bound_exists, &bound_weight, &bound_cost);
+
+	if(!bound_exists) {
+		if(single_pair_ctx->path_count == 0) {
+			single_pair_ctx->array = arr_new(WeightedPath, 0);
+		} else if(single_pair_ctx->path_count == 1) {
+			single_pair_ctx->single.path = NULL;
+		} else {
+			single_pair_ctx->heap = Heap_new(path_cmp, NULL);
+		}
```
Comment (DvirDukhan) [inline src/procedures/proc_sp_paths.c]: can you explain this logic?

File: src/procedures/proc_sp_paths.c [hunk 4]
```diff
@@ -683,20 +1135,76 @@ static ProcedureResult Proc_SPpathsInvoke
...
+	}
+
+	bool cost_feasible = (bound_cost <= single_pair_ctx->max_cost);
+	double initial_bound = cost_feasible ? bound_weight : DBL_MAX;
 
 	if(single_pair_ctx->path_count == 0) {
-		SPpaths_all_minimal(single_pair_ctx);
+		// all-minimal wants every tie at the true minimum weight; the
+		// pre-pass weight is a valid upper bound on that minimum (any tie
+		// is by definition <= it), so seeding is safe here.
```
Comment (DvirDukhan) [inline src/procedures/proc_sp_paths.c:1223]: when will it not be 0 after calling _find_bound_path?

Comment (swilly22) [reply src/procedures/proc_sp_paths.c:1223]: `find_bound_path` is a quick cheap initial search to see if `A` and `B` are connected.
if the procedure found a path between them then its cost could serve as an upper bound to any found path weight.

File: src/procedures/proc_sp_paths.c (hunk 3 above)
Comment (swilly22) [reply src/procedures/proc_sp_paths.c]: `A` can't reach `B`.

Depending on the user configuration:

1. `path_count == 0` all shortest paths
2. `path_count == 1` a single shortest path.
3. `path_count > 1` `N` shortest paths.

We need to set the shortest path context for a reply, e.g. `single_pair_ctx->array = arr_new (WeightedPath, 0) ;` no shortest paths been found.

File: src/procedures/proc_sp_paths.c (hunk 1 above)
Comment (swilly22) [reply src/procedures/proc_sp_paths.c:488]: No when searching for a shortest path between `A` and `B` the definition of a path requires at least one edge. As such we set `exists` to `false`

**Review by DvirDukhan — APPROVED** [review]
